lib/ansible/module_utils/basic.py

```diff
@@ -2020,15 +2031,15 @@ def run_command(self, args, check_rc=False, close_fds=True, executable=None, dat
         # create a printable version of the command for use
         # in reporting later, which strips out things like
         # passwords from the args list
-        if isinstance(args, basestring):
-            if isinstance(args, unicode):
-                b_args = args.encode('utf-8')
-            else:
-                b_args = args
-            to_clean_args = shlex.split(b_args)
-            del b_args
+        to_clean_args = args
+        if PY2:
+            if isinstance(args, text_type):
+                to_clean_args = args.encode('utf-8')
         else:
-            to_clean_args = args
+            if isinstance(args, binary_type):
+                to_clean_args = args.decode('utf-8', errors='replace')
+        if isinstance(args, (text_type, binary_type)):
+            to_clean_args = shlex.split(to_clean_args)
```

Comment:
Thinking about whether this should be a utility function... the problem I see is that if we attempt to hide the complexity here we come out with an unclean API.  If we don't hide the complexity inside of hte function, then the function really doesn't do anything.

What I mean by unclean API is that the function would say "Takes either text or utf-8 encoded byte strings on input.  On output gives text type on python3 and bytes type on python2".  The unrelated matrix elements make it unintuitive how the function behaves.  It would be better if the return type was linked to the input type or the input type had to be appropriate for the version of python (or if both values were constant).